pkg/certificateprovider/azurekeyvault/auth.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/confidential"
+	"github.com/pkg/errors"
+)
+
+const (
+	// the format for expires_on in UTC with AM/PM
+	expiresOnDateFormatPM = "1/2/2006 15:04:05 PM +00:00"
+	// the format for expires_on in UTC without AM/PM
+	expiresOnDateFormat = "1/2/2006 15:04:05 +00:00"
+
+	tokenTypeBearer = "Bearer"
+	// For Azure AD Workload Identity, the audience recommended for use is
+	// "api://AzureADTokenExchange"
+	DefaultTokenAudience = "api://AzureADTokenExchange" //nolint
+)
+
+// authResult contains the subset of results from token acquisition operation in ConfidentialClientApplication
+// For details see https://aka.ms/msal-net-authenticationresult
+type authResult struct {
+	accessToken    string
+	expiresOn      time.Time
+	grantedScopes  []string
+	declinedScopes []string
+}
+
+func getAuthorizerForWorkloadIdentity(ctx context.Context, tenantID, clientID, resource, aadEndpoin string) (autorest.Authorizer, error) {
+
+	scope := resource
+	// .default needs to be added to the scope
+	if !strings.Contains(resource, ".default") {
+		scope = fmt.Sprintf("%s/.default", resource)
+	}
+	result, err := getAADAccessToken(ctx, tenantID, clientID, scope)
+	if err != nil {
+		return nil, fmt.Errorf("failed to acquire token: %w", err)
+	}
+
+	token := adal.Token{
+		AccessToken: result.AccessToken,
+		Resource:    resource,
+		Type:        tokenTypeBearer,
+	}
+	token.ExpiresOn, err = parseExpiresOn(result.ExpiresOn.UTC().Local().Format(expiresOnDateFormat))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse expires_on: %w", err)
+	}
+
+	return autorest.NewBearerAuthorizer(authResult{
+		accessToken:    result.AccessToken,
+		expiresOn:      result.ExpiresOn,
+		grantedScopes:  result.GrantedScopes,
+		declinedScopes: result.DeclinedScopes,
+	}), nil
+}
+
+// OAuthToken implements the OAuthTokenProvider interface.  It returns the current access token.
+func (ar authResult) OAuthToken() string {
+	return ar.accessToken
+}
+
+// Vendored from https://github.com/Azure/go-autorest/blob/def88ef859fb980eff240c755a70597bc9b490d0/autorest/adal/token.go
+// converts expires_on to the number of seconds
+func parseExpiresOn(s string) (json.Number, error) {
+	// convert the expiration date to the number of seconds from now
+	timeToDuration := func(t time.Time) json.Number {
+		dur := t.Sub(time.Now().UTC())
+		return json.Number(strconv.FormatInt(int64(dur.Round(time.Second).Seconds()), 10))
+	}
+	if _, err := strconv.ParseInt(s, 10, 64); err == nil {
+		// this is the number of seconds case, no conversion required
+		return json.Number(s), nil
+	} else if eo, err := time.Parse(expiresOnDateFormatPM, s); err == nil {
+		return timeToDuration(eo), nil
+	} else if eo, err := time.Parse(expiresOnDateFormat, s); err == nil {
+		return timeToDuration(eo), nil
+	} else {
+		// unknown format
+		return json.Number(""), err
+	}
+}
```

Comment:
Need unit tests for this

Reply:
thanks @aramase . this reminded me of the quesiton i had releated to token := adal.Token{ , on line 73, it looks like the token value is not used?

Reply:
It is used with `OAuthToken`. Authorizer needs to have that interface implemented and that's what is called when trying to get a token.